src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,141 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> prevNode = tail;
+        Node<T> newNode = new Node<>(prevNode, value, null);
+        tail = newNode;
+        if (isEmpty()) {
+            head = newNode;
+        } else {
+            prevNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> currentNode = getNode(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        if (currentNode.prev == null) {
+            head = newNode;
+        } else {
+            currentNode.prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listItem : list) {
+            add(listItem);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = getNode(index);
+        T oldValue = oldNode.value;
+        oldNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = getNode(index);
+        unlink(node);
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (areValuesEqual(object, currentNode.value)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+        if (index < (size / 2)) {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            for (int i = index; i > 0; i--) {
+                currentNode = currentNode.next;
+            }
+        }
```

Comment:
you are iterating only from the start to the end of the nodes list because only `currentNode.next` is being called not `previous`, and start is always from the head, but the key to optimize performance if we know that index is closer to an end - we can start iteration from the tail to the index